### DISCUSSION TITLE
Remove duplicate link

### DISCUSSION THREAD

**Discussion by jarx88** [karakeep-app/karakeep#651] [Ideas]
Hello,

I imported a very large number of bookmarks from the browser. It would be helpful to have a tool that allows me to remove duplicate links.

![image|300](https://github.com/user-attachments/assets/a9ae1717-0411-43b8-975e-ca04ec973669)

**Comment by kamtschatka** [comment]
what exactly did you do? there is already link deduplication built-in

**Comment by cecoates** [comment]
@kamtschatka something that https://github.com/karakeep-app/karakeep/issues/859 mentions is that https:// and http:// will be considered as unique links.

Similarly, a trailing / is considered a unique URL.

Migrating from Pocket and Raindrop, I had quite a bit of overlap between the services at times (many readers will allow you to sync to multiple services, I figured it was a safe bet in case one went under, and it turns out that was a fair bet.)

I didn't realize it until after, but now I have hundreds of duplicate links, but no practical way to remove them or search for them.

An optional, perhaps more aggressive, way to deduplicate would be excellent! Matching on titles that are exactly the same, or URLs that are the same even if one is http and another is https, or if one has a trailing slash.

Maybe if it was possible, if content was the same, similar to how OCR is done? Like if 95% of all the text is identical?

![Screenshot 2025-06-19 at 20-02-23 Karakeep](https://github.com/user-attachments/assets/650d78f0-ae1b-4e6c-a744-dad74ee4d865)

**Reply by thiswillbeyourgithub** [reply]
If you know how to code (or vibe code) it's fairly easy to do using  the [karakeep python api](https://github.com/thiswillbeyourgithub/karakeep_python_api) and would be a nice [community script](https://github.com/thiswillbeyourgithub/karakeep_python_api/tree/main/community_scripts)! And if you don't, I suggest creating an issue with a bounty to make one.

**Reply by cecoates** [reply]
@thiswillbeyourgithub I don't know what vibe coding is, but maybe I wish I had seen your suggestion like 2 hours ago, haha.

I went ahead and cludged together a low-tech solution I posted below before I realized you replied! But I agree if someone made a script, that'd be excellent.

**Comment by cecoates** [comment]
I'm not a programmer, but to remove these duplicates here is what I did.

First, I exported my bookmarks from Karakeep to HTML.

Then, I opened that HTML file in Notepad++

Next, I did a find and replace for http:// and replaced them all with https://

After that, I did a search for /" and replaced them with "

Notepad++ also has something under "Edit" and "Line Operations", that can remove identical lines. But theoretically maybe you don't have to do that, since now the URLs should have all the trailing slashes and http:// removed.

I deleted my Karakeep instance (I wasn't sure if stuff gets left behind, and I just starting using it, so I didn't mind starting over after I backed up the data folder just in case.)

Started a new one, and imported my deduplicated HTML file. 

The original export had around 4,100 bookmarks. My deduplicated one had around 3,400.

During the import, I could see it had detected additional URL duplicates, and was importing 2,908 bookmarks.

I'm not sure what happened, but the first time I tried this, Karakeep only had 2,227 when it finished, despite the importer originally saying 2,908. I'm not clear what the hiccup was, as I walked away while it was working. 

So I imported the file again, let it run again, and this time it said importing 2,908, and when it finished, I had 2,908.

Just for kicks, to double-check everything worked out, I did some find/replace on the "deduplicated" HTML so I could leave just the URLs.

I used a regex find/replace:

`" ADD_DATE=".*`

replaced with nothing. This left just a lot of lines with:

`    <DT><A HREF="`

and the URLs. I used the duplicate line removal operation in Notepad++ again, removed the extra parts of the HTML at the beginning that weren't just the bookmark lines and... what do you know. 2,908 entries. 

This might be inefficient, but that's about as technical as I can handle. :-)

And if you've done something like full-page archive on a bunch of stuff and downloaded videos, it doesn't account for trying to save that data. But it did help me take my library of URLs from over 4,000 to under 3,000 (I guess a lot of what I had saved to Instapaper and Pocket must've overlapped).

Maybe that'll help someone else interested in cleaning up their library.

**Comment by Eragos** [comment]
Hey!

If the URI is the same it won't be saved a second time. If it is different, you get a second entry with the other URI. And that is okay.

For Example:
- https://example.com/test
- https://example.com/test#chapter1
- https://example.com/test#chapter2

Will result in hoarding of three different pages (section of pages). All have the same base. This might be in most cases the correct behaviour IMHO.

Just my2ct

Best, Michael
